diff --git a/CMakeLists.txt b/CMakeLists.txt
index e6b9933ce..da75456ad 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -359,9 +359,9 @@ option(Z3_BUILD_LIBZ3_SHARED "Build libz3 as a shared library if true, otherwise
 ################################################################################
 # Symbol visibility
 ################################################################################
-if (NOT MSVC)
-  z3_add_cxx_flag("-fvisibility=hidden" REQUIRED)
-endif()
+# if (NOT MSVC)
+#   z3_add_cxx_flag("-fvisibility=hidden" REQUIRED)
+# endif()
 
 ################################################################################
 # Tracing
